client/src/quic_client.rs

```diff
@@ -0,0 +1,229 @@
+                Ok(())
+            }
+        }
+    }
+}
+
+impl QuicClient {
+    pub fn new(client_socket: UdpSocket, addr: SocketAddr) -> Self {
+        let runtime = tokio::runtime::Builder::new_multi_thread()
+            .enable_all()
+            .build()
+            .unwrap();
+
+        let _guard = runtime.enter();
+
+        let crypto = rustls::ClientConfig::builder()
+            .with_safe_defaults()
+            .with_custom_certificate_verifier(SkipServerVerification::new())
+            .with_no_client_auth();
+
+        let create_endpoint = QuicClient::create_endpoint(EndpointConfig::default(), client_socket);
+
+        let mut endpoint = runtime.block_on(create_endpoint);
+
+        endpoint.set_default_client_config(ClientConfig::new(Arc::new(crypto)));
+
+        Self {
+            runtime,
+            endpoint,
+            connection: Arc::new(Mutex::new(None)),
+            addr,
+        }
+    }
+
+    async fn create_endpoint(config: EndpointConfig, client_socket: UdpSocket) -> Endpoint {
+        quinn::Endpoint::new(config, None, client_socket).unwrap().0
+    }
+
+    async fn _send_buffer_using_conn(
+        data: &[u8],
+        connection: &Arc<NewConnection>,
+    ) -> Result<(), WriteError> {
+        let mut send_stream = connection.connection.open_uni().await?;
+        send_stream.write_all(data).await?;
+        send_stream.finish().await?;
+        Ok(())
+    }
+
+    // Attempts to send data, connecting/reconnecting as necessary
+    // On success, returns the connection used to successfully send the data
+    pub async fn send_buffer(&self, data: &[u8]) -> Result<Arc<NewConnection>, WriteError> {
+        let connection = {
+            let mut conn_guard = self.connection.lock().await;
+
+            let maybe_conn = (*conn_guard).clone();
+            match maybe_conn {
+                Some(conn) => conn.clone(),
+                None => {
+                    let connecting = self.endpoint.connect(self.addr, "connect").unwrap();
+                    let connection = Arc::new(connecting.await?);
+                    *conn_guard = Some(connection.clone());
+                    connection
+                }
+            }
+        };
+        match Self::_send_buffer_using_conn(data, &connection).await {
+            Ok(()) => Ok(connection),
+            _ => {
+                let connection = {
+                    let connecting = self.endpoint.connect(self.addr, "connect").unwrap();
+                    let connection = Arc::new(connecting.await?);
+                    let mut conn_guard = self.connection.lock().await;
+                    *conn_guard = Some(connection.clone());
+                    connection
+                };
+                Self::_send_buffer_using_conn(data, &connection).await?;
+                Ok(connection)
+            }
+        }
+    }
+
+    pub async fn send_batch(&self, buffers: Vec<&[u8]>) -> Result<(), WriteError> {
```

Comment:
Seems like this should be returning at least a `Vec<Result<(), WriteError>` so the caller can decide what to do in error cases

Reply:
We can do this in a follow-up I think. The current consumers don't need the specific error.